backend-rust/src/transaction_reject.rs

```diff
@@ -525,322 +544,432 @@ pub struct PoolClosed {
+                let reject = TransactionRejectReason::RejectedReceive(prepared.process(tx).await?);
+                serde_json::to_value(reject)?
+            }
+        };
+        Ok(reason)
+    }
+}
+
+/// Reject receive which require processing the contract update message
+/// using the smart contract module schema before insertion.
+pub struct PreparedRejectedReceive {
+    reject_reason:    i32,
+    contract_address: ContractAddress,
+    receive_name:     String,
+    message_as_hex:   String,
+}
+
+impl PreparedRejectedReceive {
+    pub async fn process(
+        &self,
+        tx: &mut sqlx::Transaction<'static, sqlx::Postgres>,
+    ) -> anyhow::Result<RejectedReceive> {
+        // Handle and store errors
+        let (message, message_parsing_error) = match self.process_message(tx).await {
+            Ok(message) => (message, None),
+            Err(error) => (None, Some(error.to_string())),
+        };
+
+        Ok(RejectedReceive {
+            reject_reason: self.reject_reason,
+            contract_address: self.contract_address,
+            receive_name: self.receive_name.clone(),
+            message_as_hex: self.message_as_hex.clone(),
+            message,
+            message_parsing_error,
+        })
+    }
+
+    /// Parse the message using the smart contract module of the smart contract
+    /// instance.
+    async fn process_message(
+        &self,
+        tx: &mut sqlx::Transaction<'static, sqlx::Postgres>,
+    ) -> anyhow::Result<Option<String>> {
+        if self.message_as_hex.is_empty() {
+            return Ok(None);
         }
+        let schema = sqlx::query_scalar!(
+            "SELECT
+                schema
+            FROM contracts
+                JOIN smart_contract_modules
+                    ON smart_contract_modules.module_reference = contracts.module_reference
+            WHERE index = $1 AND sub_index = $2",
+            i64::try_from(self.contract_address.index.0)?,
+            i64::try_from(self.contract_address.sub_index.0)?
+        )
+        .fetch_one(tx.as_mut())
+        .await?;
+        let Some(schema) = schema else {
+            // No schema found in the smart contract module.
+            return Ok(None);
+        };
+        let schema = VersionedModuleSchema::new(&schema, &None)
+            .context("Failed to parse smart contract module schema")?;
+        let receive_name = ReceiveName::new(&self.receive_name)
+            .context("Invalid receive name for RejectedReceive")?;
+        let schema_type = match schema.get_receive_param_schema(
+            receive_name.contract_name(),
+            receive_name.entrypoint_name().into(),
+        ) {
+            Ok(t) => t,
+            Err(VersionedSchemaError::NoContractInModule)
+            | Err(VersionedSchemaError::NoReceiveInContract) => return Ok(None),
+            Err(err) => {
+                anyhow::bail!("Database bytes should be a valid VersionedModuleSchema: {}", err);
```

Comment:
This error is an `InternalError` in my opinion since the indexer should not insert an invalid schema into the database in the first place. I suggest that the function `process_message` returns the tuple (message, message_parsing_error) and returns an error on `InternalError`.